### PR TITLE
Added vlc recording feature

### PR DESCRIPTION
**Pull request by alr2413**
Here is the list of changes in this PR, 
1) Updated pubspec dependency
2) Fixed local audio/subtitle issue on android
3) added recording feature of vlc player (which is also mentioned in #227, #165)

### CODE REVIEW

File: flutter_vlc_player/android/src/main/java/software/solid/fluttervlcplayer/FlutterVlcPlayer.java [hunk 1]
```diff
@@ -465,10 +468,7 @@ long getAudioDelay() {
     }
 
     void addAudioTrack(String url, boolean isNetworkUrl, boolean isSelected) {
-        if (isNetworkUrl)
-            mediaPlayer.addSlave(Media.Slave.Type.Audio, Uri.parse(url), isSelected);
-        else
-            mediaPlayer.addSlave(Media.Slave.Type.Audio, url, isSelected);
+        mediaPlayer.addSlave(Media.Slave.Type.Audio, Uri.parse(url), isSelected);
```
Comment (solid-yuriiprykhodko) [inline flutter_vlc_player/android/src/main/java/software/solid/fluttervlcplayer/FlutterVlcPlayer.java:471]: Might as well remove the unnecessary `isNetworkUrl` argument above.

**Review by solid-yuriiprykhodko — CHANGES_REQUESTED** [review]
Great addition!

The Android implementation works good for me.

However, we need to address several problems before we can merge:
1. The recorded file is saved to the filesystem without doing any UI outputs -- testing the new logic is only possible using platform-specific IDEs (file browsing with Android Studio or downloading and exploring application container in XCode). Possibly we could add the recorded video to the list of example videos at the bottom of the main screen?
2. Given that I started recording the video, when the video has finished playing -- I can't stop the recording via the UI, the button gets stuck. We should probably automatically stop recording when the video elapses.
3. On iOS, I can only record the audio output: 
[vlc-record-2021-04-24-20h00m58s-sample.mp4-.mp4.zip](https://github.com/solid-software/flutter_vlc_player/files/6594420/vlc-record-2021-04-24-20h00m58s-sample.mp4-.mp4.zip)

File: flutter_vlc_player/lib/src/vlc_player_controller.dart [hunk 2]
```diff
@@ -893,6 +901,24 @@ class VlcPlayerController extends ValueNotifier<VlcPlayerValue> {
...
+  Future<bool?> stopRecording() async {
+    _throwIfNotInitialized('stopRecording');
+    return await _record(null);
+  }
+
+  /// Returns true if media is start/stop recording.
+  Future<bool?> _record(String? saveDirectory) async {
+    _throwIfNotInitialized('addAudioTrack');
+    return await vlcPlayerPlatform.record(_viewId, saveDirectory);
+  }
```
[the remark this answers is not in the document]
Comment (alr2413) [reply flutter_vlc_player/lib/src/vlc_player_controller.dart:920]: Ok.
i replaced the record method with stopRecording & startRecording methods and also modified the corresponding controller methods.

File: flutter_vlc_player_platform_interface/lib/src/method_channel/method_channel_vlc_player.dart [hunk 3]
```diff
@@ -512,4 +519,12 @@ class MethodChannelVlcPlayer extends VlcPlayerPlatform {
       },
     );
   }
+
+  @override
+  Future<bool?> record(int textureId, String? saveDirectory) async {
+    var response = await _api.record(RecordMessage()
```
[the remark this answers is not in the document]
Comment (alr2413) [reply flutter_vlc_player_platform_interface/lib/src/method_channel/method_channel_vlc_player.dart:525]: to make the type definition similar to the other variables defined in other methods of this class, i don't think we should change it.

File: flutter_vlc_player/android/src/main/java/software/solid/fluttervlcplayer/FlutterVlcPlayer.java (hunk 1 above)
Comment (alr2413) [reply flutter_vlc_player/android/src/main/java/software/solid/fluttervlcplayer/FlutterVlcPlayer.java:471]: fixed

**Comment by alr2413** [comment]
> Great addition!
> 
> The Android implementation works good for me.
> 
> However, we need to address several problems before we can merge:
> 
> 1. The recorded file is saved to the filesystem without doing any UI outputs -- testing the new logic is only possible using platform-specific IDEs (file browsing with Android Studio or downloading and exploring application container in XCode). Possibly we could add the recorded video to the list of example videos at the bottom of the main screen?
> 2. Given that I started recording the video, when the video has finished playing -- I can't stop the recording via the UI, the button gets stuck. We should probably automatically stop recording when the video elapses.
> 3. On iOS, I can only record the audio output:
>    [vlc-record-2021-04-24-20h00m58s-sample.mp4-.mp4.zip](https://github.com/solid-software/flutter_vlc_player/files/6594420/vlc-record-2021-04-24-20h00m58s-sample.mp4-.mp4.zip)

#1 case --->  i added the recorded video to the end of list.
#2 case ---> there is a problem that when the playing video reaches to the end, the recorded path is empty!
#3 case ---> pending... 


if you've any suggestion for #2, please share it with me.
thanks.

File: flutter_vlc_player/ios/Classes/VlcViewController.swift [hunk 4]
```diff
@@ -533,6 +542,26 @@ class VLCPlayerEventStreamHandler: NSObject, FlutterStreamHandler, VLCMediaPlaye
         }
     }
     
+    func mediaPlayerStartedRecording(_ player: VLCMediaPlayer!) {
```
[the remark this answers is not in the document]
Comment (alr2413) [reply flutter_vlc_player/ios/Classes/VlcViewController.swift:545]: i'm not professional in swift, but there is a `nonnull` annotation.
might @mitchross help us on this case.

[the remark this answers is not in the document]
Comment (mitchross) [reply flutter_vlc_player/ios/Classes/VlcViewController.swift:545]: pinging @amadeu01 too who helped in past.

**Comment by alr2413** [comment]
> case 2 ---> there is a problem that when the playing video reaches to the end, the recorded path is empty!

i've found the source of mentioned issue. if you do not stop the recording process manually and the playing video reaches to end, the vlc RecordChanged event never gets called and as a result, the record file path remains empty.

**Comment by mitchross** [comment]
@solid-software @solid-pavloprykhodko 

Anything left to be reviewed on this?

**Comment by solid-software** [comment]
@mitchross if all the requested changes are completed - please mark for re-review:

image

**Comment by solid-yuriiprykhodko** [comment]
> > case 2 ---> there is a problem that when the playing video reaches to the end, the recorded path is empty!
> 
> i've found the source of mentioned issue. if you do not stop the recording process manually and the playing video reaches to end, the vlc RecordChanged event never gets called and as a result, the record file path remains empty.

I'm thinking it might be possible to add a workaround, but I'm not entirely sure about how much complexity it will add to the vlc_player_controller. 

Generally if the video is finite, we know its duration. This means we can estimate the moment when we can stop recording, and make a call to the underlying library. The recording might be a split-second shorter than the actual end of the video, but it seems to be a reasonable tradeoff compared to adding a buggy feature.

@alr2413 Could you experiment with this?

**Comment by solid-yuriiprykhodko** [comment]
> case 1 ---> i added the recorded video to the end of list.

This feature doesn't seem to work for me, on both platforms.
When I toggle recording, the video doesn't get added to the list at the bottom.

**Comment by alr2413** [comment]
> > > case 2 ---> there is a problem that when the playing video reaches to the end, the recorded path is empty!
> > 
> > 
> > i've found the source of mentioned issue. if you do not stop the recording process manually and the playing video reaches to end, the vlc RecordChanged event never gets called and as a result, the record file path remains empty.
> 
> I'm thinking it might be possible to add a workaround, but I'm not entirely sure about how much complexity it will add to the vlc_player_controller.
> 
> Generally if the video is finite, we know its duration. This means we can estimate the moment when we can stop recording, and make a call to the underlying library. The recording might be a split-second shorter than the actual end of the video, but it seems to be a reasonable tradeoff compared to adding a buggy feature.
> 
> @alr2413 Could you experiment with this?

I don't think we should fix the vlc bug because it depends on the type of source video. Even on finite length videos i've seen some times the video gets stopped before it reaches to the end duration (you can check that with the first video on the example code). On online streaming videos, this situation is different because sometimes the end duration is always zero and never changes, sometimes it gets increased over the time and .....

Thus, i guess instead of fixing this bug we should make a note on readme about this behavior of vlc recording feature, and lets developers implement a workaround based on their use-case.

**Comment by alr2413** [comment]
> > case 1 ---> i added the recorded video to the end of list.
> 
> This feature doesn't seem to work for me, on both platforms.
> When I toggle recording, the video doesn't get added to the list at the bottom.

I don't have that issue on my real android & ios device.
@mitchross can you test it?

**Comment by quanghd96** [comment]
I need this function

**Comment by zileyuan** [comment]
I need this function

**Comment by alr2413** [comment]
ping @solid-yuriiprykhodko.

**Comment by solid-software** [comment]
Hey @alr2413, pong! We are planning to give it another look in the coming days.

File: flutter_vlc_player/example/lib/single_tab.dart [hunk 5]
```diff
@@ -127,7 +127,26 @@ class _SingleTabState extends State<SingleTab> {
...
+          child: VlcPlayerWithControls(
+            key: _key,
+            controller: _controller,
+            onStopRecording: (recordPath) {
+              listVideos.add(VideoData(
+                name: 'Recorded Video',
+                path: recordPath,
+                type: VideoType.recorded,
+              ));
+              //
```
Comment (solid-yuriiprykhodko) [inline flutter_vlc_player/example/lib/single_tab.dart:139]: Please don't use empty comments as spacers -- just a newline works well with `dartfmt`.
```suggestion

```

File: flutter_vlc_player/example/lib/single_tab.dart [hunk 6]
```diff
@@ -127,7 +127,26 @@ class _SingleTabState extends State<SingleTab> {
...
-          child: VlcPlayerWithControls(key: _key, controller: _controller),
+          child: VlcPlayerWithControls(
+            key: _key,
+            controller: _controller,
+            onStopRecording: (recordPath) {
+              listVideos.add(VideoData(
+                name: 'Recorded Video',
+                path: recordPath,
+                type: VideoType.recorded,
+              ));
```
Comment (solid-yuriiprykhodko) [inline flutter_vlc_player/example/lib/single_tab.dart:138]: Please move this inside the `setState` callback -- this way it will be clearer which piece of state is changing.

Reference: https://api.flutter.dev/flutter/widgets/State/setState.html
> Whenever you change the internal state of a State object, make the change in a function that you pass to setState:

File: flutter_vlc_player/example/lib/single_tab.dart [hunk 7]
```diff
@@ -172,6 +195,8 @@ class _SingleTabState extends State<SingleTab> {
                 ),
               ),
               onTap: () async {
+                await _controller.stopRecording();
+                //
```
Comment (solid-yuriiprykhodko) [inline flutter_vlc_player/example/lib/single_tab.dart:199]:
```suggestion

```

File: flutter_vlc_player/example/lib/single_tab.dart [hunk 8]
```diff
@@ -127,7 +127,26 @@ class _SingleTabState extends State<SingleTab> {
...
+                type: VideoType.recorded,
+              ));
+              //
+              ScaffoldMessenger.of(context).showSnackBar(
+                SnackBar(
+                  content: Text(
+                      'The recorded video file has been added to the end of list.'),
+                ),
+              );
+              //
```
Comment (solid-yuriiprykhodko) [inline flutter_vlc_player/example/lib/single_tab.dart:146]:
```suggestion

```

File: flutter_vlc_player/example/lib/vlc_player_with_controls.dart [hunk 9]
```diff
@@ -1,19 +1,23 @@
...
 import 'package:flutter/rendering.dart';
 import 'package:flutter_vlc_player/flutter_vlc_player.dart';
+import 'package:path_provider/path_provider.dart';
 
 import 'controls_overlay.dart';
 
 class VlcPlayerWithControls extends StatefulWidget {
   final VlcPlayerController controller;
   final bool showControls;
+  final Function(String) onStopRecording;
```
Comment (solid-yuriiprykhodko) [inline flutter_vlc_player/example/lib/vlc_player_with_controls.dart:14]: Let's extract this callback into a typedef, and specify a return type. Right now it is inferred as `dynamic`, and it's unclear what exactly is returned.

File: flutter_vlc_player/example/lib/vlc_player_with_controls.dart [hunk 10]
```diff
@@ -41,6 +45,11 @@ class VlcPlayerWithControlsState extends State<VlcPlayerWithControls>
   int numberOfAudioTracks = 0;
   bool validPosition = false;
 
+  //
```
Comment (solid-yuriiprykhodko) [inline flutter_vlc_player/example/lib/vlc_player_with_controls.dart:48]:
```suggestion

```

File: flutter_vlc_player/example/lib/vlc_player_with_controls.dart [hunk 11]
```diff
@@ -84,7 +93,25 @@ class VlcPlayerWithControlsState extends State<VlcPlayerWithControls>
...
+        }
+      } else {
+        recordingTextOpacity = 0;
+      }
+      // check for change in recording state
+      if (isRecording != _controller.value.isRecording) {
+        isRecording = _controller.value.isRecording;
+        if (!isRecording) {
+          if (widget.onStopRecording != null) {
+            print(_controller.value.recordPath);
```
Comment (solid-yuriiprykhodko) [inline flutter_vlc_player/example/lib/vlc_player_with_controls.dart:110]: Avoid `print`s.
```suggestion

```

File: flutter_vlc_player/ios/Classes/VlcViewBuilder.swift [hunk 12]
```diff
@@ -2,7 +2,7 @@ import Foundation
 import Flutter
 
 public class VLCViewBuilder: NSObject, VlcPlayerApi{
-    
+
```
Comment (solid-yuriiprykhodko) [inline flutter_vlc_player/ios/Classes/VlcViewBuilder.swift:5]: Unnecessary whitespace:
```suggestion
```

**Review by solid-yuriiprykhodko — CHANGES_REQUESTED** [review]
Just a few more comments and we're good to go.

I agree, let's add a note to our README describing the issue with iOS screen recording. Since this is a problem with the underlying library, not the wrapper we're supporting, it's OK to release this feature, and hopefully `libvlc` authors will fix it sometime.

**Comment by alr2413** [comment]
We fixed all the requested changes.
I tested  on my android device and it works without problem.
Waiting for @mitchross test result on IOS.

File: flutter_vlc_player/README.md [hunk 13]
```diff
@@ -146,9 +146,10 @@ class _MyHomePageState extends State<MyHomePage> {
 ```
 
 
+### Recording feature
+To start/stop video recording, you have to call the `startRecording(String saveDirectory)` and `stopRecording()` methods, respectively. By calling the stop method you can get the path of recorded file from `vlcPlayerController.value.recordPath`.
 
-
-
+<b>Note</b>: there is a bug in vlc library that if you do not call the `stopRecording()` method and the video reaches to end the recorded video path is not getting updated, because the vlc stop recording event never gets called.
```
Comment (solid-yuriiprykhodko) [inline flutter_vlc_player/README.md:152]: Let's create a `Known Issues` section closer to the bottom of the readme, and move it there.

Also, specify that:
1. The issue is specific to iOS -- I don't think I ran into the problem on Android.
2. The issue has to do with the underlying library itself, not the Dart wrapper we're supporting.

**Review by solid-yuriiprykhodko — CHANGES_REQUESTED** [review]

**Comment by mitchross** [comment]
> We fixed all the requested changes.
> I tested on my android device and it works without problem.
> Waiting for @mitchross test result on IOS.

Its working on IOS for me!

File: flutter_vlc_player/README.md (hunk 13 above)
Comment (alr2413) [reply flutter_vlc_player/README.md]: @mitchross and I tested on a real android device and the same problem occurs. So it's not specific to IOS platform.
Since we currently support both Android & IOS platforms, i don't think we need to mention it in readme.

File: flutter_vlc_player_platform_interface/lib/src/utils/options/vlc_http_options.dart [hunk 14]
```diff
@@ -0,0 +1,36 @@
+class VlcHttpOptions {
```
Comment (solid-yuriiprykhodko) [inline flutter_vlc_player_platform_interface/lib/src/utils/options/vlc_http_options.dart:1]: These changes look good, but are out of scope for the current PR. 
Please open a new PR and move these changes there. Some rationale would be useful, too.

File: flutter_vlc_player/README.md [hunk 15]
```diff
@@ -176,12 +208,23 @@ If you have some changes made to the iOS app, recreate the app using above metho
...
 ### Breaking Changes (from V4 to V5)
 Entire platform has been refactored in v5. It will require a refactor of your app to follow v5. 
 
+<br>
 <hr>
+<br>
+
+
+## Known Issues
+<b>1)</b> there is an issue in the <u>recording feature of underlying vlc library (not the Dart wrapper we're supporting)</u>, that if you do not call the `stopRecording()` method after `startRecording(...)` method and the video reaches to end the recorded video path is not getting updated, because the vlc stop recording event never gets fired.
```
Comment (solid-yuriiprykhodko) [inline flutter_vlc_player/README.md:217]: Let's rephrase a bit:

```suggestion
<b>1)</b> The video recording feature is problematic in iOS: if the video reaches its end while you're recording it, the underlying `vlckit` library fails to finalise the recording process, and we cannot retrieve the recorded file. The issue is reported and tracked here: [https://code.videolan.org/videolan/VLCKit/-/issues/394](https://code.videolan.org/videolan/VLCKit/-/issues/394) (see last comment from September 22, 2020)
```

**Review by solid-yuriiprykhodko — CHANGES_REQUESTED** [review]

File: flutter_vlc_player/README.md (hunk 15 above)
Comment (mitchross) [reply flutter_vlc_player/README.md:217]: The issue is prevalent on Android too. I was able to reproduce on Android.

Comment (solid-yuriiprykhodko) [reply flutter_vlc_player/README.md:217]: Thanks for pointing that out! 
Curiously, it didn't reproduce on my device -- maybe the problem is device-specific.

In any case, I think it would be good to leave a link for the tracked issue on VLC Gitlab. Given that the problem exists on both platforms, I suspect that the bug should be in `libvlc` per se, not its iOS/Android wrappers. Could you please try finding the issuetracker on their Gitlab? 
If not, I'll be able to make a quick search on Monday.

Comment (alr2413) [reply flutter_vlc_player/README.md]: i updated readme and i also searched about this issue on libvlc-android. 
I only found the following link:
[https://code.videolan.org/videolan/vlc-android/-/issues/2117](https://code.videolan.org/videolan/vlc-android/-/issues/2117)

Comment (solid-yuriiprykhodko) [reply flutter_vlc_player/README.md]: My thought was too look for similar issues on the Gitlab repo of the core `libvlc` library: https://code.videolan.org/videolan/vlc/-/issues.

Comment (alr2413) [reply flutter_vlc_player/README.md]: I looked at that repo too, but i didn't find any issue.

**Review by solid-yuriiprykhodko — APPROVED** [review]
Did some digging myself -- VLC repo is quite messy, and they have a separate list for bugs/issues with no shipping date, some are there for up to 15 years!

I think we're good to merge this feature, with just a note on its instability.

cc @solid-software

**Comment by mitchross** [comment]
> Did some digging myself -- VLC repo is quite messy, and they have a separate list for bugs/issues with no shipping date, some are there for up to 15 years!
> 
> I think we're good to merge this feature, with just a note on its instability.
> 
> cc @solid-software

Tell me about it. I waited 1 month for a merged bug fix to make it thru their ci/cd to cocoapods.

**Comment by solid-software** [comment]
Thanks everybody, @alr2413, @mitchross and @solid-yuriiprykhodko !
